src/schema/builder.rs

```diff
@@ -0,0 +1,592 @@
+            let value = self.types.get(&name).unwrap();
+            if &other != value {
+                let msg = format!(
+                    "Type `{}` already defined as `{:?}` but being redefined as `{:?}`",
+                    self.basic_name(&name),
+                    other,
+                    value
+                );
+                return Err(BuilderError::new(msg));
+            }
+        }
+        Ok(name)
+    }
+
+    fn name_ref(&mut self, name: &str, namespace: Option<&str>) -> Result<NameRef, BuilderError> {
+        let (name, namespace) = match name.rfind(".") {
+            Some(pos) => {
+                if name.len() == pos {
+                    return Err(BuilderError::new(format!("Invalid name `{}`", name)));
+                }
+                (&name[pos + 1..name.len()], Some(&name[0..pos]))
+            }
+            None => (name, namespace),
+        };
+
+        Ok(NameRef {
+            name: self.type_names.get_or_intern(name),
+            namespace: namespace.map(|ns| self.namespace_names.get_or_intern(ns)),
+        })
+    }
+
+    fn name_ref_alised(
+        &mut self,
+        name: &str,
+        namespace: Option<&str>,
+        aliases: Vec<&str>,
+    ) -> Result<NameRef, BuilderError> {
+        let name_ref = self.name_ref(name, namespace)?;
+
+        for alias in aliases {
+            let alias_name = self.name_ref(alias, namespace)?;
+
+            // Deal with the all important forward mapping
+            if let Some(existing) = self.aliases.insert(alias_name, name_ref) {
+                let msg = format!(
+                    "Multiple names found for alias `{}` known: `{}` vs given: `{}`",
+                    self.basic_name(&alias_name),
+                    self.basic_name(&existing),
+                    self.basic_name(&alias_name)
+                );
+                return Err(BuilderError::new(msg));
+            }
+
+            // Now deal with the reverse mapping
+            self.reverse_aliases
+                .entry(name_ref)
+                .or_default()
+                .push(alias_name);
+        }
+
+        Ok(name_ref)
+    }
+
+    fn basic_name(&self, nameref: &NameRef) -> String {
+        let name = self.type_names.resolve(nameref.name).unwrap();
+        match nameref
+            .namespace
+            .and_then(|x| self.namespace_names.resolve(x))
+        {
+            Some(ns) => format!("{}.{}", ns, name),
+            None => name.to_string(),
+        }
+    }
+}
+
+#[cfg(test)]
+mod test {
+
+    use super::*;
+
+    #[test]
+    fn test_basic_builder() -> Result<(), Error> {
+        let mut builder = SchemaBuilder::new();
+        let mut record_builder = builder.record("test");
+        record_builder.field("field1", builder.int());
+        record_builder.field("field2", builder.long());
+        let record = record_builder.build(&mut builder)?;
+
+        let schema = builder.build(record)?;
```

Comment:
I fear that this asymmetry with primitive types could actually cause a bit of confusion:
```
let builder = Schema::builder();
let root = builder.long();
builder.build(root).unwrap()
```

I understand that same kind of difference for named types must exist, but this line
```
let record = record_builder.build(&mut builder)?;
```
where the `builder` now becomes the argument of the finalizing function for the `RecordBuilder` is really twisting my brain.

The code here is pretty complex and I must admit I am struggling a bit to follow, but:
1. Is there any way to make the record type be added to the schema builder with the `record()` call and then return a mutable reference to it?
2. If not, what would be your opinion on reversing the finalizing call into something like `builder.add(record_builder)` instead of `record_builder.build(&mut bulder)`?

Reply:
Yeah the builder isn't my favorite part and I only cope with its API as I wrote it.

I will try to take another swing at it